File: scripts/quick_search.lua
local flib_table = require("__flib__.table")
local gui = require("__flib__.gui")
local on_tick_n = require("__flib__.on-tick-n")

local search_utils = require("scripts.search")
local qs_math = require("scripts.quick_search.qs_math")


local QuickSearch = {}
local QuickSearch_mt = {__index = QuickSearch}

local min_chars_for_search = 2
local dicts_to_search = {"item", "fluid", "technology"}


function QuickSearch:new(player_index)
    local o = {}
    setmetatable(o, QuickSearch_mt)

    o.player_index = player_index

    o.visible = false
    o.refs = {}
    o.search_results = {}
    o.search_has_math = false

    o.rebuild_gui = false

    return o
end

function QuickSearch:load()
    setmetatable(self, QuickSearch_mt)
    return self
end

function QuickSearch:destroy()
    if self.refs ~= nil and self.refs.frame ~= nil and type(self.refs.frame.destroy) == "function" then
        self.refs.frame.destroy()
    end
end

function QuickSearch:build_gui()
    if self.rebuild_gui == true then
        if self.refs.frame ~= nil and self.refs.frame.valid == true then
            self.refs.frame.destroy()
            self.refs = {}
        end
    end

    local player = game.get_player(self.player_index)

    if self.refs.frame == nil or self.refs.frame.valid == false then
        self.rebuild_gui = false
        self.refs = gui.build(player.gui.screen, {
            {
                type = "flow",
                direction = "vertical",
                style = "fcodex_quick_search",
                ref = {"frame"},
                actions = {
                    on_closed = "qs_close"
                },
                {type = "label", caption="QUICK SEARCH", style= "fcodex_quick_search_label", ref = {"caption"}},
                {type ="textfield", style= "fcodex_quick_search_input", ref = {"search_field"},
                 actions = {
                     on_text_changed = "qs_update_search",
                     on_confirmed = "qs_test_debug"
                 }},
                {type = "list-box", style= "fcodex_quick_search_results", ref = {"results"},
                 actions = {
                     on_selection_state_changed = "qs_try_open_codex"
                 }}
            }
        })

        self.refs.frame.visible = self.visible
    end

    self.refs.frame.location = {
        x=(player.display_resolution.width/2)-200,
        y=(player.display_resolution.height/2)-50,
    }
    self.refs.search_field.clear_and_focus_on_right_click = true
    self.refs.results.style.maximal_height = player.display_resolution.height/2 - 100
end

function QuickSearch:open()
    if not self.visible then
        self.visible = true

        self:build_gui()
        self.refs.frame.visible = true

        game.get_player(self.player_index).opened = self.refs.frame

        self.refs.frame.bring_to_front()
        self.refs.search_field.select_all()
        self.refs.search_field.focus()
    end
end

function QuickSearch:close()
    if self.visible then
        self.visible = false

        if self.refs.frame ~= nil then
            self.refs.frame.visible = false
        end

        local player = game.get_player(self.player_index)
        if player.opened then
            player.opened = nil
        end
    end
end

function QuickSearch:toggle()
    if self.visible then
        self:close()
    else
        self:open()
    end
end

function QuickSearch:is_open()
    return self.visible
end

function QuickSearch:set_rebuild_gui()
    self.rebuild_gui = true
end

function QuickSearch:display_result_list(unfiltered_list)
    self:remove_search_results()
    self.search_results = {}

    if #unfiltered_list == 0 then
        return
    end

    for _,data in ipairs(unfiltered_list) do
        if data.prototype ~= nil then
            local icon = "["..data.type.."="..data.id.."]"
            local text = data.name

            local debug_postfix = nil

            if debug:is_enabled() then
                debug_postfix = " M: "..data.match_count .. " O: " .. data.prototype.order
            end

            if data.type == "item" and data.prototype.has_flag("hidden") then
                text = nil
            elseif data.type == "technology" then
                text = "[color=#add8e6] " .. text .. "[/color]"
            end

            if text ~= nil and data.prototype.valid then
                table.insert(self.search_results, data)
                self.refs.results.add_item(icon .. text .. (debug_postfix or ""))
            end
        end
    end
end

function QuickSearch:remove_search_results()
    local math_text = nil
    if self.search_has_math == true and #self.refs.results.items >= 1 then
        math_text = self.refs.results.get_item(1)
    end

    self.refs.results.clear_items()

    if self.search_has_math == true then
        if math_text == nil then
            self.search_has_math = false
        else
            self.refs.results.add_item(math_text)
        end
    end
end

function QuickSearch:set_math_result(result, error)
    if result ~= nil or (error ~= nil and error ~= "") then
        local math_text = result ~= nil and ("=" .. result) or "=?"

        if self.search_has_math == true and #self.refs.results.items >= 1 then
            self.refs.results.set_item(1, math_text)
        else
            self.refs.results.add_item(math_text, 1)
        end

        self.math_result = result
        self.search_has_math = true
    else
        if self.search_has_math == true and #self.refs.results.items >= 1 then
            self.refs.results.remove_item(1)
        end

        self.math_result = nil
        self.search_has_math = false
    end
end

function QuickSearch:update_input(prompt)
    if self.visible == false then
        --log("Quick search not visible for player "..self.player_index)
        return
    end
    if prompt == nil then
        prompt = self.refs.search_field.text
    else
        log("Setting quick search text: "..prompt)
        self.refs.search_field.text = prompt
    end

    --log("Update input: \"".. (prompt or "") .. "\"")

    if prompt == "" then
        self.refs.results.clear_items()
        self.math_result = nil
        self.search_has_math = false

        --log("Quick search has empty prompt for player "..self.player_index)
        return
    end

    local math_prompt = string.gsub((prompt == nil) and "" or prompt, "%s+", "")

    local math_result, math_err = qs_math.calculate_result(math_prompt)
    self:set_math_result(math_result, math_err)

    local dicts_cache = global.cache:get_player_cache(self.player_index, "dicts_cache")
    if dicts_cache:is_translated() == false then
        self:remove_search_results()
        self.refs.results.add_item({ "factorio-codex.waiting-for-translation" })
        return
    end

    if self.last_search_task ~= nil then
        on_tick_n.remove(self.last_search_task)
    end
    self.last_search_task = on_tick_n.add(game.tick + 1, {
        player_index = self.player_index,
        gui = "qs",
        name = "update_search",
        prompt = prompt
    })
end

function QuickSearch:execute_task(task)
    if task.name == "update_search" then
        local dicts_cache = global.cache:get_player_cache(self.player_index, "dicts_cache")

        if dicts_cache:is_translated() == false then
            return
        end

        local dicts = {}
        for _,dict_name in pairs(dicts_to_search) do
            local dict = dicts_cache:get_names_dict(dict_name)
            if dict ~= nil then
                dicts[dict_name] = dict
            else
                log("Dict "..dict_name.." missing")
            end
        end

        local matching_names = search_utils.find(task.prompt, dicts)

        search_utils.sort(matching_names, {
            search_utils.sort_orders.hidden_last,
            search_utils.sort_orders.tech_last,
            search_utils.sort_orders.match_count,
            search_utils.sort_orders.factorio})

        --log(serpent.block(self.refs.results.valid, {maxnum=15}))

        self:display_result_list(matching_names)


        -- informatron magic stuff
        --[[for interface, functions in pairs(remote.interfaces) do
            if functions["informatron_menu"] and functions["informatron_page_content"] then

            end
        end]]

        return
    end
end

function QuickSearch:gui_action(action, event)
    if event.player_index ~= self.player_index then
        log("Error: Event received for quick search but player indexes mismatch! (QS: "..self.player_index..", E: "..event.player_index..")")
        local rec = game.get_player(self.player_index)
        local ex_rec = game.get_player(event.player_index)
        debug:player_print(self.player_index, "[factorio-codex] Error: You received an event that was for "..ex_rec.name..
                "! (Maybe invalid player data?)")

        debug:player_print(event.player_index, "[factorio-codex] Error: Your event got mistakenly sent to "..rec.name..
                "! (Maybe invalid player data?)")
    end

    local action_list = {
        qs_close = function() self:close() end,
        qs_update_search = function (event) self:update_input() end,
        qs_try_open_codex =
            function (event)
                local selected_index = event.element.selected_index

                if self.search_has_math then
                    if selected_index == 1 then
                        -- put math result into input if it is a valid formula
                        if self.math_result ~= nil then
                            self.refs.search_field.text = "" .. self.math_result
                            self.refs.search_field.focus()
                        end

                        return
                    end
                    selected_index = selected_index - 1
                end

                if  self.search_results == nil then
                    log("Cant open codex - empty result list")
                    return
                end

                --log("Quick Search result array: " .. serpent.block(quick_search.search_results))

                local selected = self.search_results[selected_index]

                if selected == nil or selected.type == "error" then
                    return
                end

                event.element.selected_index = 0

                --game.print("Index " .. event.element.selected_index .. " with content [" .. selected.type .. "=" .. selected.id .. "] \"" .. selected.name .. "\" was selected!")
                PlayerData:get_codex(event.player_index):show_info(selected.id, selected.type)
            end,
        qs_test_debug =
            function (event)
                if event.element.text == "debug!" then
                    debug:toggle(event.player_index)
                end
            end
    }

    local action_func = action_list[action]
    if action_func ~= nil then
        action_func(event)
    else
        log("Unknown action \"" .. action .. "\" for quick search!")
    end
end

function QuickSearch:validate(expected_player_indx)
    log("Quick Search Validate: Checking quick search...")
    local valid = true
    local fixed = true

    if getmetatable(self) ~= QuickSearch_mt then
        log("Quick Search Validate: Metatable is not the Quick Search metatable!")
        setmetatable(self, QuickSearch_mt)
        valid = false
        fixed = fixed and true
    end

    if self.player_index ~= expected_player_indx then
        log("Quick Search Validate: player index is wrong! E: "..expected_player_indx.." A: " .. serpent.line(self.player_index))
        self.player_index = expected_player_indx
        valid = false
        fixed = fixed and true
    end

    local not_nil_values = {
        "refs", "search_results", "visible", "search_has_math", "player_index"
    }
    local nil_detected = false

    for _,v in pairs(not_nil_values) do
        if self[v] == nil then
            log("Quick Search Validate: Detected nil value \""..v.."\" for player "..expected_player_indx)
            nil_detected = true
        end
    end

    if nil_detected then
        log("Quick Search Validate: Contents: " .. serpent.line(self, {nocode=true}))
        valid = false
        fixed = false
    end

    if self.refs.search_results ~= nil and self.refs.search_results.valid then
        local res_gui_length = #self.refs.search_results.items
        local res_length = #self.search_results

        local mismatch = res_gui_length ~= res_length

        if mismatch == true then
            log("Quick Search Validate: Mismatch between search_results gui element [1](len: "..serpent.line(res_gui_length)..
                    ") and search_results list [2](len: "..serpent.line(res_length).."):\n[1] "..
                    serpent.line(self.refs.search_results.items, {nocode=true}).."\n[2] "..
                    serpent.line(self.search_results, {nocode=true}))
            valid = false
            fixed = false
        end
    end

    return valid,fixed
end

return QuickSearch

File: src/QuickSearch.ts
import {default as Util, validate_print_info, validate_status} from "Util";
import {getDictionaryCache} from "cache/DictionaryCache";
import Search, {SortOrderQS} from "search/Search";
import QSMath from "quick_search/QS_math";
import {Task, TaskExecutor} from "Task";
import PlayerData from "PlayerData";
import IGuiRoot, {GuiAction} from "IGuiRoot";

/** @noResolution */
import * as FLIB_on_tick_n from "__flib__.on-tick-n";
/** @noResolution */
import * as FLIB_gui from "__flib__.gui";
import {getSettingsCache} from "./cache/SettingsCache";

declare const global: {
    playerData: typeof PlayerData
}

type QSResult = {
    type: string
    name: string
    id: string

    match_count: number
    hidden: boolean,
    order: string
}

const dicts_to_search = ["item", "fluid", "technology"]
const gui_name = "quick_search"

class QuickSearch implements TaskExecutor, IGuiRoot {
    player_index: PlayerIndex;
    visible: boolean;
    refs: {
        frame?: FlowGuiElement
        caption?: LabelGuiElement
        search_field?: TextFieldGuiElement
        results?: ListBoxGuiElement,
        debug?: LuaTable<string, BaseGuiElement>
    };

    rebuild_gui: boolean;


    last_search_task?: FLIBTaskIdent;

    search_results: QSResult[];
    search_has_math: boolean;
    math_result?: number;

    constructor(player_index: PlayerIndex) {
        this.player_index = player_index

        this.visible = false
        this.refs = {}
        this.rebuild_gui = false

        this.last_search_task = undefined
        this.search_results = []
        this.search_has_math = false
        this.math_result = undefined
    }

    static load(this: void, qs: QuickSearch) {
        // @ts-ignore
        setmetatable(qs, QuickSearch.prototype)
    }

    destroy() {
        this.refs?.frame?.destroy()
        this.refs = {}
    }

    build_gui() {
        if (this.rebuild_gui) {
            this.destroy()
        }

        let player = game.get_player(this.player_index)
        if (player == null) {
            $log_err!(`Error player_index is invalid! No player with index ${this.player_index} exists!`)
            return
        }

        if ( this.refs?.frame?.valid != true ) {
            this.rebuild_gui = false

            this.refs = FLIB_gui.build(player.gui.screen, [{
                type: "flow",
                direction: "vertical",
                style: "fcodex_quick_search",
                ref: ["frame"],
                actions: {
                    on_closed: { gui: gui_name, action: "close" }
                },

                1: {type: "label"    , style: "fcodex_quick_search_label", ref: ["caption"]     , caption: "QUICK SEARCH"},
                2: {type: "textfield", style: "fcodex_quick_search_input", ref: ["search_field"],
                    actions: {
                        on_text_changed: { gui: gui_name, action: "update_search" },
                        on_confirmed: { gui: gui_name, action: "test_debug" }
                    }},
                3: {type: "list-box" , style: "fcodex_quick_search_results", ref: ["results"],
                    actions: {
                        on_selection_state_changed: { gui: gui_name, action: "try_open_codex" }
                    }
                }
            }])

            if ( this.refs.frame != undefined ) this.refs.frame.visible = this.visible
        }

        if ( this.refs.search_field != undefined ) this.refs.search_field.clear_and_focus_on_right_click = true
        this.adjust_size_and_position()
    }

    adjust_size_and_position() {
        let player = game.get_player(this.player_index)

        if (player != null && this.refs.frame != undefined && this.refs.results != undefined) {
            let display_scale = player.display_scale
            let half_height = player.display_resolution.height / 2
            let half_width  = player.display_resolution.width / 2

            this.refs.frame.location = {
                x: half_width - 200 * display_scale,
                y: half_height - 50 * display_scale,
            }

            this.refs.results.style.maximal_height = (half_height - 100) / display_scale
        }
    }

    open() {
        let player = game.get_player(this.player_index)
        if (this.visible || player == null) {
            return
        }

        this.visible = true

        this.build_gui()
        if( this.refs.frame != undefined && this.refs.results != undefined && this.refs.search_field != undefined) {
            this.refs.frame.visible = true
            player.opened = this.refs.frame

            this.adjust_size_and_position()
            this.refs.frame.bring_to_front()
            this.refs.search_field.focus()
            this.refs.search_field.select_all()
        }
    }

    close() {
        if (!this.visible) {
            return
        }

        this.visible = false

        if ( this.refs.frame != undefined ) {
            this.refs.frame.visible = false
        }

        let player = game.get_player(this.player_index)
        if (player?.opened == this.refs.frame && player?.opened != undefined) {
            player.opened = undefined
        }
    }

    toggle() {
        this.visible ? this.close() : this.open();
    }

    is_open() {
        return this.visible
    }

    set_rebuild_gui() {
        this.rebuild_gui = true
    }

    display_result_list(unfiltered_list: QSResult[]) {
        this.remove_search_results()
        this.search_results = []

        if (unfiltered_list.length == 0) {
            return
        }

        for ( let data of unfiltered_list ) {
            if (data.hidden) {
                continue
            }

            let text = data.name
            switch (data.type) {
                case "technology":
                    text = `[color=#add8e6]${text}[/color]`
            }

            /*if ( debug.is_enabled() ) {
                text += ` M: ${data.match_count} O: ${data.prototype.order}`
            }*/

            this.search_results.push(data)

            if ( this.refs.results != undefined ) {
                this.refs.results.add_item(`[${data.type}=${data.id}] ${text}`)
            }
        }
    }

    remove_search_results() {
        let math_text = undefined
        if ( this.search_has_math && this.refs?.results?.items?.length != undefined && this.refs.results.items.length > 0) {
            math_text = this.refs.results.get_item(1)
        }

        if ( this.refs.results != undefined ) this.refs.results.clear_items()
        this.search_results = []

        if ( math_text != undefined ) {
            if ( this.refs.results != undefined ) this.refs.results.add_item(math_text)
        } else if( this.search_has_math ) {
            this.search_has_math = false
        }
    }

    set_math_result(result?: number, err?: string) {
        if ( result == undefined && (err == undefined || err == "") ) {
            if ( this.search_has_math && this.refs.results != undefined && this.refs.results.items.length > 0 ) {
                this.refs.results.remove_item(1)
            }

            this.math_result = undefined
            this.search_has_math = false
        } else {
            let math_text = result == undefined ? "=?" : "="+result

            if ( this.refs.results != undefined ) {
                if ( this.search_has_math && this.refs.results.items.length > 0 ) {
                    this.refs.results.set_item(1, math_text)
                } else {
                    this.refs.results.add_item(math_text, 1)
                }
            }

            this.math_result = result
            this.search_has_math = true
        }
    }

    update_input(prompt?: string) {
        if ( !this.is_open() ) return

        if ( prompt == undefined ) {
            prompt = this.refs?.search_field?.text
        } else if ( this.refs.search_field != undefined ) {
            this.refs.search_field.text = prompt
        }

        if ( prompt == undefined || prompt == "" ) {
            if ( this.refs.results != undefined ) {
                this.refs.results.clear_items()
            }

            this.math_result = undefined
            this.search_has_math = false
            return;
        }

        let math_prompt = prompt.replace(" ", "")
        let [math_result, math_err] = QSMath.calculateString(math_prompt)
        this.set_math_result(math_result, math_err)

        if (this.last_search_task != undefined) {
            FLIB_on_tick_n.remove(this.last_search_task)
        }
        let task: Task = {
            type: "gui",
            player_index: this.player_index,
            gui: "qs",
            args: new LuaTable<string, any>()
        }
        task.args?.set("name", "update_search")
        task.args?.set("prompt", prompt)

        this.last_search_task = FLIB_on_tick_n.add(game.tick + 1, task)
    }

    execute_task(task: Task) {
        if (task.args?.get("name") == "update_search") {
            let dictsCache = getDictionaryCache(this.player_index)
            if (dictsCache == undefined) {
                $log_warn!("Could not retrieve Dictionary cache!")
                return
            }
            if ( !dictsCache.isTranslated() ) {
                this.remove_search_results()
                this.refs.results?.add_item([ "factorio-codex.waiting-for-translation" ])
                return;
            }

            let matching_names = Search.search(task.args?.get("prompt"), this.player_index, [
                SortOrderQS.hidden_last,
                SortOrderQS.tech_last,
                SortOrderQS.match_count,
                SortOrderQS.factorio],
                100)
            //$log_info!(serpent.block(matching_names,  {}))

            this.display_result_list(matching_names)
        }
    }

    gui_action(guiAction: GuiAction, event: GuiEventData) {
        let action = guiAction.action
        if ( event.player_index != this.player_index ) {
            $log_err!("Something is not right. Received event for another player! "+
            `Expected player id: ${this.player_index} got: ${event.player_index}!`)
            game.get_player(event.player_index)?.print("[factorio-codex] Your UI interaction could not be processed! Is mod data corrupted?")
            return
        }


        if (action == "update_search") {
            this.update_input()
        } else if (action == "close") {
            this.close()
        } else if ( action == "test_debug" ) {
            if (event.element?.text == "debug!") {
                getSettingsCache(this.player_index)?.toggleDebug()
            }
        } else if ( action == "try_open_codex" ) {
            let selectedIndex = event.element?.selected_index as uint
            if (selectedIndex == undefined || selectedIndex == 0) return

            if (selectedIndex == 1 && this.search_has_math && this.math_result != undefined) {
                if ( this.refs.search_field != undefined) this.refs.search_field.text = "" + this.math_result
                return;
            }

            selectedIndex -= 1 + (this.search_has_math ? 1 : 0)

            if (this.search_results.length < selectedIndex) {
                $log_info!("Aborted opening Codex: Index out of range!"+
                `Selected element: ${selectedIndex} but only ${this.search_results.length} elements are available!`)
                return;
            }
            let selectedResult = this.search_results[selectedIndex]

            if (selectedResult == undefined) {
                $log_info!(`Aborted opening Codex: Selected element is undefined! Index: ${selectedIndex}`)
                return;
            }

            // @ts-ignore
            if (event.element?.selected_index != undefined) event.element.selected_index = 0


            global.playerData.getCodex(event.player_index)?.show_info(selectedResult.id, selectedResult.type)
        } else if ( action == "debugToggle") {
            if (this.refs.debug == undefined) return;

            let  is_debug = getSettingsCache(this.player_index)?.is_debug()
            is_debug = is_debug ?? false

            for (let [id, debugElem] of this.refs.debug) {
                if (debugElem != undefined) {
                    debugElem.visible = is_debug
                }
            }
        } else {
            $log_info!(`Received Unknown action "${action}" for QuickSearch of player ${this.player_index
            }('${game.get_player(this.player_index)?.name}')`)
        }
    }

    validate(print_info: validate_print_info, indx: PlayerIndex): validate_status {
        let status: validate_status = validate_status.OK
        let curStatus

        curStatus = validate_status.OK
        if ( this.player_index != indx ) {
            this.player_index = indx
            curStatus = validate_status.FIXED
        }
        $log_info!(Util.format_validate_msg(print_info, "player_index", curStatus))
        if (status == validate_status.OK) status = curStatus;
        else if(status == validate_status.FIXED) status = curStatus != validate_status.OK ? curStatus : status;


        curStatus = validate_status.OK
        if ( this.visible == undefined ) {
            curStatus = validate_status.ERROR
        }
        $log_info!(Util.format_validate_msg(print_info, "visible", curStatus))
        if (status == validate_status.OK) status = curStatus;
        else if(status == validate_status.FIXED) status = curStatus != validate_status.OK ? curStatus : status;

        curStatus = validate_status.OK
        if ( this.refs == undefined) {
            this.refs = {}
            curStatus = validate_status.FIXED
        }
        $log_info!(Util.format_validate_msg(print_info, "refs", curStatus))
        if (status == validate_status.OK) status = curStatus;
        else if(status == validate_status.FIXED) status = curStatus != validate_status.OK ? curStatus : status;

        curStatus = validate_status.OK
        if ( this.rebuild_gui == undefined ) {
            this.rebuild_gui = true
            curStatus = validate_status.FIXED
        }
        $log_info!(Util.format_validate_msg(print_info, "rebuild_gui", curStatus))
        if (status == validate_status.OK) status = curStatus;
        else if(status == validate_status.FIXED) status = curStatus != validate_status.OK ? curStatus : status;

        curStatus = validate_status.OK
        if ( this.search_results == undefined ) {
            this.search_results = []
            curStatus = validate_status.FIXED
        }
        $log_info!(Util.format_validate_msg(print_info, "search_results", curStatus))
        if (status == validate_status.OK) status = curStatus;
        else if(status == validate_status.FIXED) status = curStatus != validate_status.OK ? curStatus : status;

        curStatus = validate_status.OK
        if ( this.search_has_math == undefined ) {
            curStatus = validate_status.ERROR
        }
        $log_info!(Util.format_validate_msg(print_info, "search_has_math", curStatus))
        if (status == validate_status.OK) status = curStatus;
        else if(status == validate_status.FIXED) status = curStatus != validate_status.OK ? curStatus : status;

        curStatus = validate_status.OK
        if ( this.search_has_math && this.math_result == undefined  ) {
            this.search_results = []
            curStatus = validate_status.ERROR
        }
        $log_info!(Util.format_validate_msg(print_info, "math_result", curStatus))
        if (status == validate_status.OK) status = curStatus;
        else if(status == validate_status.FIXED) status = curStatus != validate_status.OK ? curStatus : status;


        return status
    }
}

export default QuickSearch